packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs

```diff
@@ -146,169 +150,144 @@ impl PostgresDatabase {
-            "#,
-        )
-        .bind(version_id)
-        .bind(
-            serde_json::to_value(property_type)
-                .unwrap_or_else(|err| unreachable!("Could not serialize property type: {err}")),
-        )
-        .bind(created_by)
-        .fetch_one(&self.pool)
-        .await
-        .report()
-        .change_context(InsertionError)
-        .attach_lazy(|| property_type.clone())?;
+    /// - If the [`BaseUri`] does not already exist
+    ///
+    /// [`BaseUri`]: crate::types::BaseUri
+    async fn update<T>(
+        &self,
+        database_type: T,
+        updated_by: AccountId,
+    ) -> Result<Qualified<T>, UpdateError>
+    where
+        T: DatabaseType + Serialize + Send + Sync,
+    {
+        let uri = database_type.uri();
 
-        Ok(())
+        if !self
+            .contains_base_uri(uri.base_uri())
+            .await
+            .change_context(UpdateError)?
+        {
+            return Err(Report::new(BaseUriDoesNotExist)
+                .attach_printable(uri.base_uri().clone())
+                .change_context(UpdateError));
+        }
+
+        let version_id = self.insert_uri(uri).await.change_context(UpdateError)?;
+
+        self.insert_with_id(version_id, &database_type, updated_by)
+            .await
+            .change_context(UpdateError)?;
+
+        Ok(Qualified::new(version_id, database_type, updated_by))
     }
 
-    /// Inserts an [`EntityType`] identified by [`VersionId`], and associated with an [`AccountId`],
-    /// into the database.
+    /// Inserts a [`DatabaseType`] identified by [`VersionId`], and associated with an
+    /// [`AccountId`], into the database.
     ///
     /// # Errors
     ///
     /// - [`DatastoreError`], if inserting failed.
-    async fn insert_entity_type(
-        &mut self,
+    async fn insert_with_id<T>(
+        &self,
         version_id: VersionId,
-        entity_type: &EntityType,
+        database_type: &T,
         created_by: AccountId,
-    ) -> Result<(), InsertionError> {
-        sqlx::query_as::<_, (Uuid,)>(
+    ) -> Result<(), InsertionError>
+    where
+        T: DatabaseType + Serialize + Sync,
+    {
+        sqlx::query(&format!(
             r#"
-            INSERT INTO entity_types (
+            INSERT INTO {} (
                 version_id,
                 schema,
                 created_by
             ) 
-            VALUES ($1, $2, $3)
-            RETURNING version_id;
+            VALUES ($1, $2, $3);
             "#,
-        )
+            T::table()
+        ))
         .bind(version_id)
         .bind(
-            serde_json::to_value(entity_type)
-                .unwrap_or_else(|err| unreachable!("Could not serialize entity type: {err}")),
+            serde_json::to_value(database_type)
+                .unwrap_or_else(|err| unreachable!("Could not serialize data type: {err}")),
         )
         .bind(created_by)
-        .fetch_one(&self.pool)
+        .execute(&self.pool)
```

Comment:
We were using `RETURNING` clauses in the insert queries with `fetch_one` to ensure that insertion didn't return 0 affected rows. Right now this query would be able to insert nothing and still be considered a successful call. Do we want that, or is inserting nothing considered an `InsertionError`? (I would lean against inserting nothing being an error personally)

It is a little unclear, so perhaps adding a comment to make the usage of returning clear would be preferable.

The alternative is checking the return value of `execute` after awaiting to see if `affected_rows` is equal to one, which would do the same as using the `RETURNING` clause with `fetch_one` for us.

Reply:
That makes sense. I added a check, that not inserting a datum is an error: 2fcbe81cbc1927dcb98a0eafbab398c8f9030f09